server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -47,234 +40,74 @@
-            );
-        }
-    }
-
-    boolean isAutoImportDanglingIndicesEnabled() {
-        return this.isAutoImportDanglingIndicesEnabled;
-    }
-
-    /**
-     * Process dangling indices based on the provided meta data, handling cleanup, finding
-     * new dangling indices, and allocating outstanding ones.
-     */
-    public void processDanglingIndices(final Metadata metadata) {
-        assert this.isAutoImportDanglingIndicesEnabled;
-
-        if (nodeEnv.hasNodeFile() == false) {
-            return;
-        }
-        cleanupAllocatedDangledIndices(metadata);
-        findNewAndAddDanglingIndices(metadata);
-        allocateDanglingIndices(metadata);
     }
 
     /**
-     * Either return the current set of dangling indices, if auto-import is enabled, otherwise
-     * scan for dangling indices right away.
+     * Finds new dangling indices by iterating over the indices and trying to find indices
+     * that have state on disk, but are not part of the provided metadata.
      * @return a map of currently-known dangling indices
      */
     public Map<Index, IndexMetadata> getDanglingIndices() {
-        if (this.isAutoImportDanglingIndicesEnabled) {
-            // This might be a good use case for CopyOnWriteHashMap
-            return Map.copyOf(this.danglingIndices);
-        } else {
-            return findNewDanglingIndices(emptyMap(), this.clusterService.state().metadata());
-        }
-    }
-
-    /**
-     * Cleans dangling indices if they are already allocated on the provided meta data.
-     */
-    void cleanupAllocatedDangledIndices(Metadata metadata) {
-        for (Index index : danglingIndices.keySet()) {
-            final IndexMetadata indexMetadata = metadata.index(index);
-            if (indexMetadata != null && indexMetadata.getIndex().getName().equals(index.getName())) {
-                if (indexMetadata.getIndex().getUUID().equals(index.getUUID()) == false) {
-                    logger.warn("[{}] can not be imported as a dangling index, as there is already another index " +
-                        "with the same name but a different uuid. local index will be ignored (but not deleted)", index);
-                } else {
-                    logger.debug("[{}] no longer dangling (created), removing from dangling list", index);
-                }
-                danglingIndices.remove(index);
-            }
-        }
+        final Metadata metadata = this.clusterService.state().metadata();
+        return findDanglingIndices(metadata);
     }
 
-    /**
-     * Finds (@{link #findNewAndAddDanglingIndices}) and adds the new dangling indices
-     * to the currently tracked dangling indices.
-     */
-    void findNewAndAddDanglingIndices(final Metadata metadata) {
-        final IndexGraveyard graveyard = metadata.indexGraveyard();
+    // Extracted from getDanglingIndices() as a package-private method to allow easier testing testing
```

Comment:
Looks like the tests already mock out the `ClusterService`, so I think we could pass in the metadata via that mock, have them call `getDanglingIndices()` and thereby inline this. Or have `DanglingIndicesState` track a `Supplier<ClusterState>` rather than the full `ClusterService` and pass that in for the tests. Not sure which I prefer really, up to you.